### PR TITLE
Username validation for special characters

### PR DESCRIPTION
**Pull request by rutuja-amazon**
Add username validation:

- [x] Username should not include special characters

### CODE REVIEW

**Comment by shikharj05** [comment]
Thank you for the change. Can you add tests as well?

**Comment by rutuja-amazon** [comment]
Added unit tests.

**Comment by cwperks** [comment]
@rutuja-amazon Should I add any backport labels to this PR?

File: src/test/java/org/opensearch/security/dlic/rest/api/UserApiTest.java [hunk 1]
```diff
@@ -133,6 +133,10 @@ public void testUserApi() throws Exception {
         response = rh.executePutRequest(ENDPOINT + "/internalusers/", "{\"hash\": \"123\"}", new Header[0]);
         Assert.assertEquals(HttpStatus.SC_METHOD_NOT_ALLOWED, response.getStatusCode());
 
+        // username has special characters
+        response = rh.executePutRequest(ENDPOINT + "/internalusers/n@ag:ilum", "{\"hash\": \"123\"}", new Header[0]);
+        Assert.assertEquals(HttpStatus.SC_METHOD_NOT_ALLOWED, response.getStatusCode());
```
Comment (cliu123) [inline src/test/java/org/opensearch/security/dlic/rest/api/UserApiTest.java:138]: Is this test case specifically related to this fix? If yes, why isn't it also `HttpStatus.SC_BAD_REQUEST`.

Comment (rutuja-amazon) [reply src/test/java/org/opensearch/security/dlic/rest/api/UserApiTest.java:138]: Yes, we need to updated all test cases that use special characters in `username` to now fail at creation

Comment (cliu123) [reply src/test/java/org/opensearch/security/dlic/rest/api/UserApiTest.java:138]: Agreed. Then shouldn't this `PUT` request with special characters in `username` get `HttpStatus.SC_BAD_REQUEST`, which would be consistent with the fix?

File: src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java [hunk 2]
```diff
@@ -93,6 +94,12 @@ protected void handlePut(RestChannel channel, final RestRequest request, final C
             return;
         }
 
+        Pattern usernamePattern = Pattern.compile("[$&+,:;=\\\\?@#|/'<>.^*()%!-]");
```
Comment (shikharj05) [inline src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java:97]: I think ``.`` should be allowed

Comment (rutuja-amazon) [reply src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java]: Permitted `.`. Do we also allow `@` in username?

File: src/test/java/org/opensearch/security/dlic/rest/api/UserApiTest.java (hunk 1 above)
Comment (rutuja-amazon) [reply src/test/java/org/opensearch/security/dlic/rest/api/UserApiTest.java]: Sounds good, updated to `HttpStatus.SC_BAD_REQUEST`.

File: src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java (hunk 2 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java]: `@` shouldn't be allowed and only one `.` should be allowed. Thoughts? @cwperks @peternied @davidlago

File: src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java [hunk 3]
```diff
@@ -93,6 +94,12 @@ protected void handlePut(RestChannel channel, final RestRequest request, final C
             return;
         }
 
+        Pattern usernamePattern = Pattern.compile("[$&+,:;=\\\\?@#|/'<>^*()%!-]");
```
Comment (peternied) [inline src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java:97]: > Username should not include special characters

Why not, are there bugs associated with these characters?

**Review by peternied — CHANGES_REQUESTED** [review]
This change impacts forward combability as username might have been viable in older versions of OpenSearch that will no longer work.

File: src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java (hunk 2 above)
Comment (rutuja-amazon) [reply src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java]: Do we also allow `-`, `_` and `@` along with `.`? Please suggest.
In the current fix I have permitted the four special characters above.

File: src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java (hunk 3 above)
Comment (peternied) [reply src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java]: Thanks for looking out for issues of this nature, could you be more specific as to the risks associated with individual characters?  I am reticent to limit these usernames without cause.

If there is an issue please follow our reporting process [1] also quoted below.

> ## Reporting a Vulnerability
>
>If you discover a potential security issue in this project we ask that you notify AWS/Amazon Security via our [vulnerability reporting page](http://aws.amazon.com/security/vulnerability-reporting/) or directly via email to aws-security@amazon.com. Please do **not** create a public GitHub issue.

[1] https://github.com/opensearch-project/security/security/policy

**Comment by rutuja-amazon** [comment]
Requesting addition of backport labels @cwperks

**Comment by DarshitChanpura** [comment]
@rutuja-amazon which branches would you like this change to be backported to?

**Comment by peternied** [comment]
@rutuja-amazon Since this is change is based on a LOW security finding we should discuss everything in the public on this pull request.  Pulling in the context from our internal threads:

> ## Bug
> ### Repro 
> Create two users through the dashboards create internal user:
>   - username: hello password: world:a_great_day
>   - username: hello:world password: a_great_day
> 
> Attempt to logging to login via basic auth into both accounts
> curl -k -s https://hello:world:a_great_day@localhost:9200/_plugins/_security/api/account | jq
> curl -k -s https://hello:world:a_great_day@localhost:9200/_plugins/_security/api/account | jq
>
> You cannot login as the user `hello:world`.
>
> ### Additional details
> Note; this is expected as the username cannot contain a colon [1].
>
> As per CVSS rating (3.1), it's been categorized as LOW risk.
>
> [1] https://stackoverflow.com/questions/11612854/http-https-basic-authentication-colon-in-username


For the moment, lets block **only** the colon character.

File: src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java (hunk 3 above)
Comment (peternied) [reply src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java]: Added more details on the main thread https://github.com/opensearch-project/security/pull/2277#issuecomment-1333993589

**Comment by peternied** [comment]
@rutuja-amazon I've created https://github.com/rutuja-amazon/security/pull/1, if you merge that pull request into this change we can get this issue fixed, what do you think?

**Comment by rutuja-amazon** [comment]
> @rutuja-amazon I've created [rutuja-amazon#1](https://github.com/rutuja-amazon/security/pull/1), if you merge that pull request into this change we can get this issue fixed, what do you think?

Issue is with other special characters as well, as discussed offline

**Comment by peternied** [comment]
I've created an issue to track the set of special characters that should be included or not, and at this time the only exclusion that seems justified is colon `:`.  Please feel free to discuss on the issue, but as this pull request stands its too restrictive without justification for me approve and merge.

- https://github.com/opensearch-project/security/issues/2309

**Review by DarshitChanpura — APPROVED** [review]

**Comment by rutuja-amazon** [comment]
Updated PR to only allow `:` for this use-case. Requesting merge of this PR with backported labels.

**Review by cwperks — COMMENTED** [review]
The changes look good to me, once the code hygiene check is fixed I will approve this PR.

**Comment by peternied** [comment]
> The changes look good to me, once the code hygiene check is fixed I will approve this PR.

I'll take care of this to help get this merged more quickly...

**Review by cwperks — APPROVED** [review]

**Comment by peternied** [comment]
Note; `CI / build-artifact-names (pull_request)` will fail because this hasn't merged the latest from main, and I'm am attempting to avoid rebasing these changes.

**Review by peternied — APPROVED** [review]

**Comment by peternied** [comment]
Failing CI check is known fixed in main, merging!

**Pull request merged**